### PR TITLE
enable M3R users to upload OWL/OBO files using importer #3209

### PR DESCRIPTION
**Pull request by jjettenn**
1. I can upload OWL file using normal uploader. The OBO filereader (The same that we are using for OWL) is not getting the correct IRI for the Ontology
 
2. Ontology/OntologyTerm data structure is optimized for queries.

3. Data API provides the ontology.

### CODE REVIEW

File: molgenis-ontology-core/src/main/java/org/molgenis/ontology/utils/NgramGenerator.java [hunk 1]
```diff
@@ -0,0 +1,274 @@
...
+		for (int i = 0; i < listOfValues.size(); i++)
+		{
+			sample[i] = listOfValues.get(i);
+		}
+
+		double tTest = TestUtils.tTest(expectedValue, sample);
+		return tTest < T_TEST_ALPHA;
+	}
+
+	public static void main(String args[]) throws OWLOntologyCreationException
```
Comment (bartcharbon) [inline molgenis-ontology-core/src/main/java/org/molgenis/ontology/utils/NgramGenerator.java]: What is the goal of this "main" function, added voor debug?
Please remove commented code

File: molgenis-data/src/main/java/org/molgenis/data/FileRepositoryCollectionFactory.java [hunk 2]
```diff
@@ -56,16 +56,12 @@ public void addFileRepositoryCollectionClass(Class<? extends FileRepositoryColle
...
-		for (Entry<String, Class<? extends FileRepositoryCollection>> entry : fileRepositoryCollection.entrySet())
-		{
-			if (name.endsWith('.' + entry.getKey()))
-			{
-				clazz = entry.getValue();
-				break;
-			}
-		}
+		
+		String extension = FileExtensionUtils.findExtansionFromSetForGenericImporter(file.getName(),
```
Comment (fdlk) [inline molgenis-data/src/main/java/org/molgenis/data/FileRepositoryCollectionFactory.java]: typo: extEnsion

File: molgenis-data/src/main/java/org/molgenis/data/importer/ImportServiceFactory.java [hunk 3]
```diff
@@ -34,11 +38,30 @@ public void addImportService(ImportService importService)
...
+					importServicesImportableMapedToExtentions.put(extension.toLowerCase(), importService);
+				}
+			}
 		}
 
-		throw new MolgenisDataException("Can not import file. No suitable importer found");
+		String extension = FileExtensionUtils.findExtansionFromSetForGenericImporter(file.getName(),
+				importServicesImportableMapedToExtentions.keySet());
+
+		final ImportService importService = importServicesImportableMapedToExtentions.get(extension);
```
Comment (fdlk) [inline molgenis-data/src/main/java/org/molgenis/data/importer/ImportServiceFactory.java]: 2 typos: mapPed to extenSions

File: molgenis-data/src/main/java/org/molgenis/data/importer/ImportServiceFactory.java [hunk 4]
```diff
@@ -34,11 +38,30 @@ public void addImportService(ImportService importService)
...
+
+		System.out.println("extension: " + extension);
```
Comment (fdlk) [inline molgenis-data/src/main/java/org/molgenis/data/importer/ImportServiceFactory.java]: sysout

File: molgenis-ontology-core/src/main/java/org/molgenis/ontology/repository/OntologyRepositoryCollection.java [hunk 5]
```diff
@@ -0,0 +1,77 @@
...
+		}
+		else
+		{
+			throw new IllegalArgumentException("Not a obo.zip or owl.zip file [" + file.getName() + "]");
+		}
+
+		try
+		{
+			List<File> uploadedFiles;
+			uploadedFiles = ZipFileUtil.unzip(file);
```
Comment (fdlk) [inline molgenis-ontology-core/src/main/java/org/molgenis/ontology/repository/OntologyRepositoryCollection.java]: might as well initialize right away then, i.e.
    List<File> uploadedFiles = ZipFileUtil.unzip(file);

File: molgenis-ontology-core/src/main/java/org/molgenis/ontology/beans/ComparableEntity.java [hunk 6]
```diff
@@ -3,38 +3,52 @@
 import java.math.BigDecimal;
 
 import org.molgenis.data.Entity;
+import org.molgenis.data.support.MapEntity;
+import org.molgenis.ontology.repository.OntologyTermQueryRepository;
 
-public class ComparableEntity implements Comparable<ComparableEntity>
+public class ComparableEntity extends MapEntity implements Comparable<ComparableEntity>
```
Comment (fdlk) [inline molgenis-ontology-core/src/main/java/org/molgenis/ontology/beans/ComparableEntity.java]: Wouldn't a Comparator<Entity> be more obvious and less intrusive than creating a whole new entity class?

File: molgenis-ontology-core/src/main/java/org/molgenis/ontology/service/OntologyServiceImpl.java [hunk 7]
```diff
@@ -244,57 +316,76 @@ else if (inputEntity.get(attributeName) != null
...
-				entityMetaData))
+		Set<String> processedOntologyTerms = new HashSet<String>();
+		Iterable<Entity> entities = searchService.search(new QueryImpl(finalQueryRule).pageSize(MAX_NUMBER_MATCHES),
+				entityMetaData);
+
+		for (Entity entity : entities)
 		{
-			BigDecimal maxNgramScore = new BigDecimal(0);
-			for (String inputAttrName : inputEntity.getAttributeNames())
+			if (!processedOntologyTerms.contains(entity.getString(OntologyTermQueryRepository.SYNONYMS)))
```
Comment (fdlk) [inline molgenis-ontology-core/src/main/java/org/molgenis/ontology/service/OntologyServiceImpl.java]: I think I'd prefer splitting this into a couple smaller methods

File: molgenis-ontology-core/src/main/java/org/molgenis/ontology/beans/ComparableEntity.java (hunk 6 above)
Comment (jjettenn) [inline molgenis-ontology-core/src/main/java/org/molgenis/ontology/beans/ComparableEntity.java]: Chao will refactor the working of this class later. So it is not necessary to make the change now

**Pull request merged**